File: 풀이/Eunsun/todo/done/[플로이드워셜]n단논법(15723).py
n = int(input()) # 전제의 수 input
graph = [[1e9 for i in range(26)] for i in range(26)] # 2차원 리스트를 생성하고 무한으로 초기화(알파벳이 26개)

for i in range(n):
    inList = input().split() # 전제 input
    start = ord(inList[0]) - ord('a') # 전제의 앞부분을 숫자 형태로 변환
    end = ord(inList[2]) - ord('a') # 전제의 뒷부분을 숫자 형태로 변환
    graph[start][end] = 1 # 해당 부분에 graph 리스트에 무한으로 되어 있는 걸 1로 변경

# 플로이드 워셜 알고리즘
for k in range(26):
    for a in range(26):
        for b in range(26):
            graph[a][b] = min(graph[a][b], graph[a][k] + graph[k][b])

m = int(input()) # 결론의 수
for i in range(m):
    inList = input().split() # 결론 input
    start = ord(inList[0]) - ord('a') # 결론의 앞부분을 숫자 형태로 변환
    end = ord(inList[2]) - ord('a') # 결론의 뒷부분을 숫자 형태로 변환
    if graph[start][end] < 1e9: # 갈 수 있는 거리라면 (값이 무한만 아니라면)
        print('T')
    else:
        print('F')

File: 풀이/Minhyeok/done/[플로이드워셜]n단논법(15723).js
// 15723

const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const n = parseInt(input[0]); // 전제의 수
const premises = input.slice(1, n + 1).map((line) => line.split(" is ")); // 전제를 분리하여 배열로 변환
const m = parseInt(input[n + 1]); // 결론의 수
const conclusions = input.slice(n + 2).map((line) => line.split(" is ")); // 결론을 분리하여 배열로 변환

let graph = {};

// 그래프 생성
for (let premise of premises) {
  const [start, end] = premise; // 각 전제에서 시작과 끝 노드 분리
  if (!graph[start]) {
    graph[start] = []; // 만약 해당 시작 노드가 그래프에 없으면 초기화
  }
  graph[start].push(end); // 그래프에 연결된 노드 추가
}

// 주어진 시작점에서 목표점까지 이르는 경로가 있는지 확인하는 DFS 함수
const dfs = (start, target, visited) => {
  if (!graph[start]) return false; // 시작점이 그래프에 없으면 바로 false 반환
  if (graph[start].includes(target)) return true; // 시작점에서 바로 목표점으로 연결되어 있다면 true 반환

  visited[start] = true; // 현재 노드를 방문했음을 표시
  for (let neighbor of graph[start]) {
    // 현재 노드와 연결된 모든 이웃 노드에 대해
    if (!visited[neighbor]) {
      // 만약 아직 방문하지 않은 이웃 노드라면
      if (dfs(neighbor, target, visited)) return true; // 해당 이웃 노드에서 목표점까지의 경로가 있는지 재귀적으로 확인
    }
  }
  return false; // 모든 경로를 확인했지만 목표점까지의 경로를 찾지 못했으면 false 반환
};

// 각 결론 확인
for (let conclusion of conclusions) {
  const [start, target] = conclusion; // 각 결론에서 시작과 목표 노드 분리
  let visited = {}; // 방문한 노드를 기록할 객체 초기화
  if (dfs(start, target, visited)) {
    // 시작 노드에서 목표 노드까지의 경로가 있다면
    console.log("T");
  } else {
    console.log("F");
  }
}
